Mage.Sets/src/mage/cards/e/EcologicalAppreciation.java

```diff
@@ -0,0 +1,150 @@
+
+    EcologicalAppreciationEffect() {
+        super(Outcome.Benefit);
+        staticText = "search your library and graveyard for up to four creature cards with different names that each have mana value X or less and reveal them. An opponent chooses two of those cards. Shuffle the chosen cards into your library and put the rest onto the battlefield";
+    }
+
+    private EcologicalAppreciationEffect(final EcologicalAppreciationEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public EcologicalAppreciationEffect copy() {
+        return new EcologicalAppreciationEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        if (player == null) {
+            return false;
+        }
+        int xValue = source.getManaCostsToPay().getX();
+        TargetCard targetCardsInLibrary = new EcologicalAppreciationTarget(Zone.LIBRARY, 4, xValue);
+        targetCardsInLibrary.setNotTarget(true);
+        boolean searched = player.choose(Outcome.PutCreatureInPlay, new CardsImpl(player.getLibrary().getCards(game)), targetCardsInLibrary, game);
+        Cards cards = new CardsImpl(targetCardsInLibrary.getTargets());
+
+        if (cards.isEmpty()) {
+            if (searched) {
+                player.shuffleLibrary(source, game);
+            }
+            return false;
+        }
+
+        int remainingCards = 4 - cards.size();
+        if (remainingCards > 0) {
+            TargetCard targetCardsInGY = new EcologicalAppreciationTarget(Zone.GRAVEYARD, remainingCards, xValue);
+            targetCardsInGY.setNotTarget(true);
+            player.choose(Outcome.PutCreatureInPlay, new CardsImpl(player.getGraveyard().getCards(game)), targetCardsInGY, game);
+            cards.addAll(targetCardsInGY.getTargets());
+        }
+
+        TargetOpponent targetOpponent = new TargetOpponent();
+        targetOpponent.setNotTarget(true);
+        player.choose(outcome, targetOpponent, source.getSourceId(), game);
+        Player opponent = game.getPlayer(targetOpponent.getFirstTarget());
+        if (opponent == null) {
+            if (searched) {
+                player.shuffleLibrary(source, game);
+            }
+            return false;
+        }
+
+        TargetCard chosenCards = new TargetCard(2, Zone.ALL, StaticFilters.FILTER_CARD);
+        chosenCards.setNotTarget(true);
+        opponent.choose(outcome, cards, chosenCards, game);
+        Cards toShuffle = new CardsImpl(chosenCards.getTargets().stream()
+            .map(game::getCard)
+            .collect(Collectors.toList()));
+
+        player.putCardsOnTopOfLibrary(toShuffle, game, source, false);
+        player.shuffleLibrary(source, game);
+        cards.removeAll(toShuffle);
+
+        player.moveCards(cards, Zone.BATTLEFIELD, source, game);
+
+        return true;
+    }
+}
+
+class EcologicalAppreciationTarget extends TargetCard {
```

Comment:
I removed the `EcologicalAppreciationTarget` class and just used `TargetCardInLibrary` and `TargetCardInYourGraveyard`. I was able to use the cmc predicate to add a filter. However, I wasn't sure how to do the different names predicate, so I extracted that logic to a method and made both targets anon classes overriding `canTarget`, which seemed to be a good compromise to avoid dupe code. Also, in the GY case, I needed to also block selecting cards that were chosen from library.

Reply:
Create new filter on apply (`filter.copy`) and add dynamic cmc and names predicates to it.

Name predicate usage example:
* `filter.add(Predicates.not(new NamePredicate("ignore name")));`

Reply:
Hmm. But I can see the problem with predicates now -- you still allows to select same names in one step, so `canTarget` is good approach.